File: backend/src/main/java/com/dominoes/lobby/service/LobbyService.java
package com.dominoes.lobby.service;

import com.dominoes.lobby.dto.LobbyStateDto;
import com.dominoes.lobby.dto.UserDto;
import com.dominoes.lobby.entity.Lobby;
import com.dominoes.lobby.entity.User;
import com.dominoes.lobby.exception.GameInProgressException;
import com.dominoes.lobby.exception.LobbyFullException;
import com.dominoes.lobby.repository.LobbyRepository;
import com.dominoes.lobby.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class LobbyService {

    private final LobbyRepository lobbyRepository;
    private final UserRepository userRepository;
    private final GameService gameService;

    @Value("${lobby.size}")
    private int lobbySize;

    @Transactional
    public synchronized User joinLobby(String nickname) {
        Lobby lobby = getOrCreateLobby();
        if (lobby.isInProgress()) {
            throw new GameInProgressException();
        }
        long currentCount = userRepository.countByLobby(lobby);
        if (currentCount >= lobby.getSize()) {
            throw new LobbyFullException();
        }

        User user = new User(nickname.trim(), lobby);
        return userRepository.save(user);
    }

    @Transactional
    public synchronized void leaveLobby(UUID userId) {
        userRepository.findById(userId).ifPresent(user -> {
            Lobby lobby = user.getLobby();
            if (lobby.isInProgress()) {
                gameService.endGame(lobby);
            }
            userRepository.deleteById(userId);
        });
    }

    @Transactional(readOnly = true)
    public LobbyStateDto getLobbyState() {
        return lobbyRepository.findFirstByOrderByIdAsc()
                .map(this::toLobbyState)
                .orElse(new LobbyStateDto(null, lobbySize, List.of()));
    }

    private Lobby getOrCreateLobby() {
        return lobbyRepository.findFirstByOrderByIdAsc()
                .orElseGet(() -> lobbyRepository.save(new Lobby(lobbySize)));
    }

    private LobbyStateDto toLobbyState(Lobby lobby) {
        List<UserDto> users = userRepository.findByLobbyOrderByNicknameAsc(lobby).stream()
                .map(user -> new UserDto(user.getId(), user.getNickname()))
                .toList();
        return new LobbyStateDto(lobby.getId(), lobby.getSize(), users);
    }
}

File: backend/internal/service/lobby_service.go
package service

import (
	"github.com/dominoes/lobby/internal/dto"
	"github.com/dominoes/lobby/internal/entity"
	"github.com/dominoes/lobby/internal/exception"
	"github.com/dominoes/lobby/internal/repository"
	"github.com/google/uuid"
)

type LobbyService struct {
	lobbyRepository *repository.LobbyRepository
	gameService     *GameService
	lobbySize       int
}

func NewLobbyService(lobbyRepository *repository.LobbyRepository, gameService *GameService, lobbySize int) *LobbyService {
	return &LobbyService{
		lobbyRepository: lobbyRepository,
		gameService:     gameService,
		lobbySize:       lobbySize,
	}
}

func (s *LobbyService) JoinLobby(nickname string) (*entity.User, error) {
	s.lobbyRepository.Lock()
	defer s.lobbyRepository.Unlock()

	lobby := s.getOrCreateLobby()
	if lobby.InProgress {
		return nil, &exception.GameInProgressError{}
	}
	if s.lobbyRepository.CountByLobby(lobby) >= lobby.Size {
		return nil, &exception.LobbyFullError{}
	}

	user := entity.NewUser(nickname, lobby)
	return s.lobbyRepository.SaveUser(user), nil
}

func (s *LobbyService) LeaveLobby(userID uuid.UUID) {
	s.lobbyRepository.Lock()
	defer s.lobbyRepository.Unlock()

	user, ok := s.lobbyRepository.FindUserByID(userID)
	if !ok {
		return
	}
	lobby, ok := s.lobbyRepository.FindFirstByOrderByIDAsc()
	if !ok {
		return
	}
	if lobby.InProgress {
		s.gameService.endGameLocked(lobby, nil)
	}
	s.gameService.OnUserLeftLobby(lobby, userID)
	s.lobbyRepository.DeleteUserByID(userID)
	_ = user
}

func (s *LobbyService) GetLobbyState() dto.LobbyStateDto {
	s.lobbyRepository.Lock()
	defer s.lobbyRepository.Unlock()

	lobby, ok := s.lobbyRepository.FindFirstByOrderByIDAsc()
	if !ok {
		return dto.LobbyStateDto{LobbyID: nil, Size: s.lobbySize, Users: []dto.UserDto{}}
	}
	return s.toLobbyState(lobby)
}

func (s *LobbyService) getOrCreateLobby() *entity.Lobby {
	lobby, ok := s.lobbyRepository.FindFirstByOrderByIDAsc()
	if ok {
		return lobby
	}
	return s.lobbyRepository.SaveLobby(entity.NewLobby(s.lobbySize))
}

func (s *LobbyService) toLobbyState(lobby *entity.Lobby) dto.LobbyStateDto {
	users := s.lobbyRepository.FindByLobbyOrderByJoinedAtAsc(lobby)
	userDtos := make([]dto.UserDto, 0, len(users))
	for _, user := range users {
		var handCount *int
		if lobby.InProgress {
			count := len(user.Hand)
			handCount = &count
		}
		userDtos = append(userDtos, dto.UserDto{
			ID:        user.ID.String(),
			Nickname:  user.Nickname,
			HandCount: handCount,
		})
	}
	lobbyID := lobby.ID
	return dto.LobbyStateDto{
		LobbyID: &lobbyID,
		Size:    lobby.Size,
		Users:   userDtos,
	}
}
